File: packages/ios/Sources/CivUI/CivDatePicker.swift
// CivUI — CivDatePicker for SwiftUI
// Accessible date picker following government design system patterns.
// Renders: label → hint → error → native DatePicker (Section 508 compliant)

import SwiftUI

/// Accessible date picker for government applications.
///
/// Wraps SwiftUI's native `DatePicker` with CivUI form field chrome:
/// label, hint, error, and token-based styling. Uses ISO date string
/// binding for compatibility with the web component API.
///
/// Usage:
/// ```swift
/// CivDatePicker(
///     label: "Appointment date",
///     value: $appointmentDate,
///     hint: "Select a date within the next 30 days",
///     min: "2026-03-18",
///     max: "2026-04-17"
/// )
/// ```
public struct CivDatePicker: View {
    // MARK: - Properties

    /// Visible label text.
    public let label: String

    /// Bound date value as ISO string (YYYY-MM-DD).
    @Binding public var value: String

    /// Help text shown below the label.
    public var hint: String?

    /// Error message.
    public var error: String?

    /// Minimum date (YYYY-MM-DD).
    public var min: String?

    /// Maximum date (YYYY-MM-DD).
    public var max: String?

    /// Placeholder text shown when no date is selected.
    public var placeholder: String

    /// Density variant — "default" or "sm" (compact). Suppresses chrome
    /// for use in dense surfaces like data-grid cell editors.
    public var spacing: String

    /// Whether a date is required.
    public var isRequired: Bool

    /// Whether the picker is disabled.
    public var isDisabled: Bool

    /// Whether the picker is read-only (visible but non-interactive, no opacity change).
    public var isReadonly: Bool

    /// Called when the date changes.
    public var onChange: ((String) -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    /// Optional form state for centralized validation.
    public var formState: CivFormState?

    /// Field name for form state registration.
    public var formName: String?

    /// Custom required message for form validation.
    public var requiredMessage: String?

    /// Custom validation function. Returns error string or nil.
    public var formValidate: (() -> String?)?

    /// Locale identifier for date formatting (e.g., "es", "fr"). When empty, the system locale is used.
    public var locale: String

    /// Whether the field contains PII (excluded from getFormData).
    public var isPii: Bool

    /// Day the week starts on (0 = Sunday, 1 = Monday).
    public var weekStartsOn: Int

    /// Comma-separated list of disabled dates (YYYY-MM-DD).
    public var disabledDates: String

    /// Label for the clear button.
    public var clearLabel: String

    /// Label for the choose-date button.
    public var chooseDateLabel: String

    /// Label announced when a date is selected.
    public var selectedDateLabel: String

    /// Accessible label for the dialog.
    public var dialogLabel: String

    /// Label for the previous month button.
    public var previousMonthLabel: String

    /// Label for the next month button.
    public var nextMonthLabel: String

    /// Message announced when the dialog is opened.
    public var dialogOpenedMessage: String

    /// Message announced when a date is selected.
    public var dateSelectedMessage: String

    /// Label for today's date in the calendar.
    public var todayLabel: String

    /// Label for the today shortcut button.
    public var todayButtonLabel: String

    /// Whether to hide the today shortcut button.
    public var hideTodayButton: Bool

    /// Whether to render an inline "Today" shortcut inset in the input chrome.
    public var inlineToday: Bool

    /// Override the inline "Today" button label.
    public var inlineTodayLabel: String

    /// Error message for invalid date format.
    public var invalidFormatMessage: String

    /// Error message for dates outside the allowed range.
    public var dateRangeMessage: String

    /// Error message for dates before the minimum.
    public var minDateMessage: String

    /// Error message for dates after the maximum.
    public var maxDateMessage: String

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        label: String,
        value: Binding<String>,
        hint: String? = nil,
        error: String? = nil,
        min: String? = nil,
        max: String? = nil,
        placeholder: String = "",
        spacing: String = "default",
        isRequired: Bool = false,
        isDisabled: Bool = false,
        isReadonly: Bool = false,
        onChange: ((String) -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil,
        formState: CivFormState? = nil,
        formName: String? = nil,
        requiredMessage: String? = nil,
        formValidate: (() -> String?)? = nil,
        locale: String = "",
        isPii: Bool = false,
        weekStartsOn: Int = 0,
        disabledDates: String = "",
        clearLabel: String = "",
        chooseDateLabel: String = "",
        selectedDateLabel: String = "",
        dialogLabel: String = "",
        previousMonthLabel: String = "",
        nextMonthLabel: String = "",
        dialogOpenedMessage: String = "",
        dateSelectedMessage: String = "",
        todayLabel: String = "",
        todayButtonLabel: String = "",
        hideTodayButton: Bool = false,
        inlineToday: Bool = false,
        inlineTodayLabel: String = "",
        invalidFormatMessage: String = "",
        dateRangeMessage: String = "",
        minDateMessage: String = "",
        maxDateMessage: String = ""
    ) {
        self.label = label
        self._value = value
        self.hint = hint
        self.error = error
        self.min = min
        self.max = max
        self.placeholder = placeholder
        self.spacing = spacing
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.isReadonly = isReadonly
        self.onChange = onChange
        self.onAnalytics = onAnalytics
        self.formState = formState
        self.formName = formName
        self.requiredMessage = requiredMessage
        self.formValidate = formValidate
        self.locale = locale
        self.isPii = isPii
        self.weekStartsOn = weekStartsOn
        self.disabledDates = disabledDates
        self.clearLabel = clearLabel
        self.chooseDateLabel = chooseDateLabel
        self.selectedDateLabel = selectedDateLabel
        self.dialogLabel = dialogLabel
        self.previousMonthLabel = previousMonthLabel
        self.nextMonthLabel = nextMonthLabel
        self.dialogOpenedMessage = dialogOpenedMessage
        self.dateSelectedMessage = dateSelectedMessage
        self.todayLabel = todayLabel
        self.todayButtonLabel = todayButtonLabel
        self.hideTodayButton = hideTodayButton
        self.inlineToday = inlineToday
        self.inlineTodayLabel = inlineTodayLabel
        self.invalidFormatMessage = invalidFormatMessage
        self.dateRangeMessage = dateRangeMessage
        self.minDateMessage = minDateMessage
        self.maxDateMessage = maxDateMessage
    }

    // MARK: - Date Helpers

    private static let isoFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    private var dateBinding: Binding<Date> {
        Binding(
            get: {
                if !value.isEmpty, let date = Self.isoFormatter.date(from: value) {
                    return date
                }
                return Date()
            },
            set: { newDate in
                let iso = Self.isoFormatter.string(from: newDate)
                value = iso
                onChange?(iso)
                onAnalytics?("change", ["value": iso])
            }
        )
    }

    private var dateRange: ClosedRange<Date>? {
        let minDate = min.flatMap { Self.isoFormatter.date(from: $0) }
        let maxDate = max.flatMap { Self.isoFormatter.date(from: $0) }

        if let minDate, let maxDate {
            return minDate...maxDate
        } else if let minDate {
            return minDate...Date.distantFuture
        } else if let maxDate {
            return Date.distantPast...maxDate
        }
        return nil
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Label
            labelView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
            }

            // 4. DatePicker or Placeholder
            if value.isEmpty && !placeholder.isEmpty {
                Text(placeholder)
                    .font(.system(size: CivTokens.Typography.FontSize.base))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .padding(.horizontal, CivTokens.Spacing._2)
                    .padding(.vertical, CivTokens.Spacing._1_5)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(adaptiveColor(
                        light: CivTokens.Colors.White.default_,
                        dark: CivTokens.DarkColors.White.default_
                    ))
                    .cornerRadius(CivTokens.Border.Radius.default_)
                    .overlay(
                        RoundedRectangle(cornerRadius: CivTokens.Border.Radius.default_)
                            .stroke(adaptiveColor(
                                light: CivTokens.Colors.Base.light,
                                dark: CivTokens.DarkColors.Base.light
                            ), lineWidth: CivTokens.Border.Width.default_)
                    )
                    .accessibilityLabel(accessibilityLabelText)
                    .accessibilityHint(accessibilityHintText)
                    .onTapGesture {
                        // Set to today to trigger the date picker
                        let today = Self.isoFormatter.string(from: Date())
                        value = today
                        onChange?(today)
                    }
            } else {
                Group {
                    if let range = dateRange {
                        DatePicker(
                            "",
                            selection: dateBinding,
                            in: range,
                            displayedComponents: .date
                        )
                        .labelsHidden()
                    } else {
                        DatePicker(
                            "",
                            selection: dateBinding,
                            displayedComponents: .date
                        )
                        .labelsHidden()
                    }
                }
                .datePickerStyle(.compact)
                .tint(adaptiveColor(
                    light: CivTokens.Colors.Primary.default_,
                    dark: CivTokens.DarkColors.Primary.default_
                ))
                .disabled(isDisabled || isReadonly)
                .opacity(isDisabled ? 0.5 : 1.0)
                .accessibilityLabel(accessibilityLabelText)
                .accessibilityHint(accessibilityHintText)
            }
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .accessibilityElement(children: .contain)
        .onChange(of: error) { newError in
            if let newError, !newError.isEmpty {
                UIAccessibility.post(notification: .announcement, argument: newError)
            }
        }
        .onAppear { registerWithFormState() }
        .onDisappear { unregisterFromFormState() }
    }

    // MARK: - Form State Registration

    private func registerWithFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.register(CivFormState.CivFieldRegistration(
            name: formName,
            label: label,
            getValue: { value },
            setValue: { _ in },
            isRequired: isRequired,
            requiredMessage: requiredMessage ?? "",
            validate: formValidate,
            isPii: isPii
        ))
    }

    private func unregisterFromFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.unregister(formName)
    }

    // MARK: - Subviews

    private var labelView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(label)
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.base,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    // MARK: - Accessibility

    private var accessibilityLabelText: String {
        var parts = [label]
        if isRequired { parts.append("required") }
        return parts.joined(separator: ", ")
    }

    private var accessibilityHintText: String {
        var parts: [String] = []
        if let hint, !hint.isEmpty { parts.append(hint) }
        if let error, !error.isEmpty { parts.append("Error: \(error)") }
        return parts.joined(separator: ". ")
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivDatePicker_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var appointment = ""
        @State private var deadline = "2026-04-01"

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivDatePicker(
                        label: "Appointment date",
                        value: $appointment,
                        hint: "Select a date for your appointment",
                        isRequired: true
                    )

                    CivDatePicker(
                        label: "Submission deadline",
                        value: $deadline,
                        max: "2026-12-31"
                    )

                    CivDatePicker(
                        label: "Disabled date",
                        value: .constant(""),
                        isDisabled: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivDatePicker.kt
// CivUI — CivDatePicker for Jetpack Compose
// Accessible date picker using Material 3 DatePickerDialog.
// Renders: label -> hint -> error -> text field with calendar button (Section 508 compliant)

package gov.civui.components

import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.DateRange
import androidx.compose.material3.DatePicker
import androidx.compose.material3.DatePickerDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.SelectableDates
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.i18n.CivLocale
import gov.civui.tokens.CivTokens
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * Accessible date picker for government applications.
 *
 * Uses Material 3 DatePickerDialog with a text input for manual entry.
 * Value is in ISO format (YYYY-MM-DD). Supports min/max date constraints.
 *
 * Usage:
 * ```kotlin
 * var date by remember { mutableStateOf("") }
 * CivDatePicker(
 *     label = "Appointment date",
 *     value = date,
 *     onValueChange = { date = it },
 *     hint = "For example: 03/15/2026",
 * )
 * ```
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CivDatePicker(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    hint: String? = null,
    error: String? = null,
    min: String? = null,
    max: String? = null,
    required: Boolean = false,
    disabled: Boolean = false,
    readonly: Boolean = false,
    placeholder: String = "MM/DD/YYYY",
    name: String = "",
    formState: CivFormState? = null,
    locale: String = "",
    weekStartsOn: Int = 0,
    disabledDates: String = "",
    clearLabel: String = "",
    chooseDateLabel: String = "",
    selectedDateLabel: String = "",
    dialogLabel: String = "",
    previousMonthLabel: String = "",
    nextMonthLabel: String = "",
    dialogOpenedMessage: String = "",
    dateSelectedMessage: String = "",
    todayLabel: String = "",
    todayButtonLabel: String = "",
    hideTodayButton: Boolean = false,
    invalidFormatMessage: String = "",
    dateRangeMessage: String = "",
    minDateMessage: String = "",
    maxDateMessage: String = "",
    onChange: ((String) -> Unit)? = null,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
) {
    val isDark = isSystemInDarkTheme()

    // Form state registration
    var formError by remember { mutableStateOf("") }
    val effectiveError = error ?: formError.ifEmpty { null }

    if (formState != null && name.isNotEmpty()) {
        androidx.compose.runtime.DisposableEffect(name) {
            formState.register(CivFormState.CivFieldRegistration(
                name = name,
                getValue = { value },
                setValue = { onValueChange(it) },
                isRequired = required,
                getError = { formError },
                setError = { formError = it },
            ))
            onDispose { formState.unregister(name) }
        }
    }

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_
    val backgroundColor = if (isDark) CivTokens.DarkColors.White.default_ else CivTokens.Colors.White.default_
    val borderColor = if (isDark) CivTokens.DarkColors.Base.light else CivTokens.Colors.Base.light

    var showDialog by remember { mutableStateOf(false) }
    var inputText by remember(value) {
        mutableStateOf(
            if (value.isNotEmpty()) {
                try {
                    val date = LocalDate.parse(value, DateTimeFormatter.ISO_LOCAL_DATE)
                    date.format(DateTimeFormatter.ofPattern("MM/dd/yyyy"))
                } catch (_: DateTimeParseException) {
                    value
                }
            } else ""
        )
    }

    // Parse min/max to epoch millis for Material DatePicker
    val minMillis = min?.let { parseIsoToEpochMillis(it) }
    val maxMillis = max?.let { parseIsoToEpochMillis(it) }

    // Parse current value to epoch millis for initial selection
    val initialMillis = if (value.isNotEmpty()) parseIsoToEpochMillis(value) else null

    Column(
        modifier = modifier.padding(bottom = CivTokens.Spacing._4),
    ) {
        // 1. Label
        CivLabel(
            label = label,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(text = hint, color = hintColor)

        // 3. Error
        CivError(text = effectiveError, color = errorColor)

        // 4. Input with calendar button
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextField(
                value = inputText,
                onValueChange = { newText ->
                    inputText = newText
                    // Try to parse the input text as a date
                    val parsed = tryParseDisplayDate(newText)
                    if (parsed != null) {
                        onValueChange(parsed)
                        onChange?.invoke(parsed)
                        onAnalytics?.invoke("change", mapOf("field" to label, "value" to parsed))
                    }
                },
                modifier = Modifier
                    .weight(1f)
                    .border(
                        width = if (effectiveError != null) CivTokens.Border.Width._2 else CivTokens.Border.Width.default_,
                        color = if (effectiveError != null) errorColor else borderColor,
                        shape = RoundedCornerShape(
                            topStart = CivTokens.Border.Radius.default_,
                            bottomStart = CivTokens.Border.Radius.default_,
                            topEnd = CivTokens.Border.Radius.none,
                            bottomEnd = CivTokens.Border.Radius.none,
                        ),
                    )
                    .alpha(if (disabled) 0.5f else 1f)
                    .semantics {
                        contentDescription = buildString {
                            append(label)
                            if (required) append(", required")
                            if (hint != null) append(". $hint")
                            if (effectiveError != null) append(". Error: $effectiveError")
                        }
                        if (effectiveError != null) {
                            error(effectiveError)
                        }
                    },
                enabled = !disabled && !readonly,
                placeholder = { Text(text = placeholder, color = hintColor) },
                textStyle = TextStyle(
                    fontSize = CivTokens.Typography.FontSize.base,
                ),
                singleLine = true,
                shape = RoundedCornerShape(
                    topStart = CivTokens.Border.Radius.default_,
                    bottomStart = CivTokens.Border.Radius.default_,
                    topEnd = CivTokens.Border.Radius.none,
                    bottomEnd = CivTokens.Border.Radius.none,
                ),
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = backgroundColor,
                    unfocusedContainerColor = backgroundColor,
                    disabledContainerColor = backgroundColor,
                    focusedIndicatorColor = Color.Transparent,
                    unfocusedIndicatorColor = Color.Transparent,
                    disabledIndicatorColor = Color.Transparent,
                ),
            )

            IconButton(
                onClick = { showDialog = true },
                enabled = !disabled && !readonly,
                modifier = Modifier
                    .border(
                        width = CivTokens.Border.Width.default_,
                        color = borderColor,
                        shape = RoundedCornerShape(
                            topStart = CivTokens.Border.Radius.none,
                            bottomStart = CivTokens.Border.Radius.none,
                            topEnd = CivTokens.Border.Radius.default_,
                            bottomEnd = CivTokens.Border.Radius.default_,
                        ),
                    )
                    .size(56.dp)
                    .alpha(if (disabled) 0.5f else 1f),
            ) {
                Icon(
                    imageVector = Icons.Default.DateRange,
                    contentDescription = CivLocale.t("datePickerChooseDateLabel"),
                )
            }
        }
    }

    // Date picker dialog
    if (showDialog) {
        val selectableDates = remember(minMillis, maxMillis) {
            object : SelectableDates {
                override fun isSelectableDate(utcTimeMillis: Long): Boolean {
                    if (minMillis != null && utcTimeMillis < minMillis) return false
                    if (maxMillis != null && utcTimeMillis > maxMillis) return false
                    return true
                }
            }
        }

        val datePickerState = rememberDatePickerState(
            initialSelectedDateMillis = initialMillis,
            selectableDates = selectableDates,
        )

        DatePickerDialog(
            onDismissRequest = { showDialog = false },
            confirmButton = {
                TextButton(
                    onClick = {
                        datePickerState.selectedDateMillis?.let { millis ->
                            val date = Instant.ofEpochMilli(millis)
                                .atZone(ZoneOffset.UTC)
                                .toLocalDate()
                            val iso = date.format(DateTimeFormatter.ISO_LOCAL_DATE)
                            onValueChange(iso)
                            onChange?.invoke(iso)
                            onAnalytics?.invoke("change", mapOf("field" to label, "value" to iso))
                            inputText = date.format(DateTimeFormatter.ofPattern("MM/dd/yyyy"))
                        }
                        showDialog = false
                    },
                ) {
                    Text("OK")
                }
            },
            dismissButton = {
                TextButton(onClick = { showDialog = false }) {
                    Text("Cancel")
                }
            },
        ) {
            DatePicker(state = datePickerState)
        }
    }
}

/**
 * Parse an ISO date string (YYYY-MM-DD) to epoch milliseconds (UTC).
 */
private fun parseIsoToEpochMillis(iso: String): Long? {
    return try {
        val date = LocalDate.parse(iso, DateTimeFormatter.ISO_LOCAL_DATE)
        date.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli()
    } catch (_: DateTimeParseException) {
        null
    }
}

/**
 * Try to parse a display date (MM/DD/YYYY) to ISO format (YYYY-MM-DD).
 */
private fun tryParseDisplayDate(text: String): String? {
    return try {
        val date = LocalDate.parse(text, DateTimeFormatter.ofPattern("MM/dd/yyyy"))
        date.format(DateTimeFormatter.ISO_LOCAL_DATE)
    } catch (_: DateTimeParseException) {
        null
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivDatePicker")
@Composable
private fun CivDatePickerPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var date by remember { mutableStateOf("") }
        CivDatePicker(
            label = "Appointment date",
            value = date,
            onValueChange = { date = it },
            hint = "For example: 03/15/2026",
            required = true,
        )

        var bounded by remember { mutableStateOf("") }
        CivDatePicker(
            label = "Date of service",
            value = bounded,
            onValueChange = { bounded = it },
            min = "2024-01-01",
            max = "2026-12-31",
        )

        CivDatePicker(
            label = "Disabled date",
            value = "",
            onValueChange = {},
            disabled = true,
        )
    }
}
